### PR TITLE
update the encryption spec to use HMAC and hash the entire header

### PR DESCRIPTION
**Pull request by oconnor663**
I ended up going with the "reserializing the header is allowed approach" rather than framing, but we can go the other way if you prefer. I didn't try to truncate the 32-byte `crypto_auth` output, though that's an option too.

@maxtaco

### CODE REVIEW

File: go/saltpack/specs/saltpack_encryption.md [hunk 1]
```diff
@@ -135,8 +139,31 @@ the same, clients may indicate that a message is "intentionally anonymous" as
...
+implementation used with saltpack has to follow these rules (most of them do):
+
+- Encodings must be minimal. For example, an array of less than 16 elements
+  must use the `fixarray` format.
+- Raw bytes must use the `bin` format. (Some implementations [require a
+  flag](https://github.com/msgpack/msgpack-python#string-and-binary-type) for
+  this.)
+- Positive integers must used the unsigned integer formats.
+
+The **MAC key** is a 32-byte secret key, unique to each recipient. It's
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: say what the nonce here is.

File: go/saltpack/specs/saltpack_encryption.md [hunk 2]
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
...
 recipients from each other.
 
-We compute each recipient's authenticator in three steps:
+We compute the authenticators in three steps:
 
-- Compute the SHA512 of the **payload secretbox**.
-- Encrypt that hash in a NaCl box, using the sender and recipient's long-term
-  keys. See [Nonces](#nonces) below.
-- The authenticator is the first 16 bytes of that box.
+- Serialize the packet number as a 64-bit unsigned big-endian integer, where
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: Just use the payload box nonce here, so 16 zeros before the 64-bit unsigned big-endian integer.

File: go/saltpack/specs/saltpack_encryption.md [hunk 3]
```diff
@@ -148,7 +175,7 @@ A payload packet is a MessagePack list with these contents:
 ]
 ```
 
-- The **authenticators list** contains 16-byte Poly1305 tags, one for each
+- The **authenticators list** contains 32-byte HMAC tags, one for each
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: they also auth the header

File: go/saltpack/specs/saltpack_encryption.md [hunk 4]
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
...
 - `"encryption nonce prefix\0"`
 - the 32-byte **ephemeral public key**
 
-The nonce for each box is then the concatenation of `P` and a 64-bit big-endian
-unsigned counter. For each **recipient box** the counter is 0. For each payload
-packet we then increment the counter, so the first packet's **authenticators
-list** uses 1, the next uses 2, and so on. For each **payload secretbox**, the
-nonce is the same as for the associated authenticators. The strict ordering of
-nonces should make it impossible to drop or reorder any payload packets.
+The 24-byte nonce for each public-key box is then concatenation of `P` and an
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: Can you call it *P* rather than `P`?

**Comment by oconnor663** [comment]
TODO: add a link to the msgpack spec

**Comment by maxtaco** [comment]
Other TODOS:
- Make a forward reference the first time you mention **payload key** since it's out of the blue.

File: go/saltpack/specs/saltpack_encryption.md [hunk 5]
```diff
@@ -160,72 +187,79 @@ recipients, then she will have the key, and she could modify the payload. The
 **authenticators list** is there to prevent this attack, protecting the
 recipients from each other.
 
-We compute each recipient's authenticator in three steps:
+We compute the authenticators in three steps:
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: Right above here, you say "protecting the recipient from each other." That used to be true, but we're doing double-duty here and also protecting the header. And this is the only we we protect the header (only the payload which is double-protected).

**Comment by maxtaco** [comment]
Another high-level comment is that the `## Header Packet` section seems to ramble a little bit now.  I wonder if it's possible to add more structure there?

**Comment by oconnor663** [comment]
I split the Header section up a bit, and added explicit instructions for the sender and recipient. Maybe easier to read at https://github.com/keybase/client/blob/jack/test/go/saltpack/specs/saltpack_encryption.md.

File: go/saltpack/specs/saltpack.md [hunk 6]
```diff
@@ -18,7 +18,7 @@ We start with some simplifying decisions:
 - We'll use [MessagePack](http://msgpack.org/index.html) for all the binary
   formatting.
 
-Thus, "SaltPack" (pronounced "NaCl pack"). The spec is in three parts:
+Thus, "saltpack" (pronounced "NaCl pack"). The spec is in three parts:
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack.md]: nuke joke, sorry!

File: go/saltpack/specs/saltpack_encryption.md [hunk 7]
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
...
+
+When composing a message, the sender follows these steps to generate the
+header:
+
+1. Generate a random 32-byte **payload key**.
+2. Generate a random ephemeral keypair, using
+   [`crypto_box_keypair`](http://nacl.cr.yp.to/box.html).
+3. Encrypt the sender's long-term public key using
+   [`crypto_secretbox`](http://nacl.cr.yp.to/secretbox.html) with the **payload
+   key**, to create the **sender secretbox**. (See [Nonces](#nonces).)
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: Instead say `(See [Nonces](#nonces) below for a description of which nonce to used)`.

File: go/saltpack/specs/saltpack_encryption.md [hunk 8]
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
...
+1. Deserialize the **header length** from the message stream using MessagePack.
+2. Read exactly **header length** bytes from the message stream.
+3. As above, hash the bytes from #2 to give the **header hash**.
+4. Deserialize the bytes from #2 using MessagePack to give the header list.
+5. Sanity check the **format name**, **version**, and **mode**.
+6. Precompute the ephemeral shared secret using
+   [`crypto_box_beforenm`](http://nacl.cr.yp.to/box.html) with the **ephemeral
+   public key** and the sender's private key.
+7. Try to open each of the **payload key boxes** in the recipients list using
+   [`crypto_box_afternm`] and the shared secret from #6. (See
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: URL for this one?

File: go/saltpack/specs/saltpack_encryption.md [hunk 9]
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
...
+10. For each recipient, encrypt 32 zero bytes using
+    [`crypto_box`](http://nacl.cr.yp.to/box.html) with the recipient's public
+    key and the sender's long-term private key. (See [Nonces](#nonces).) Take
+    the last 32 bytes of each box. These are the **MAC keys**.
+
+Encrypting the sender's long-term public key in step #3 allows Alice to stay
+anonymous to Mallory. If Alice wants to be anonymous to Bob as well, she can
+reuse the ephemeral key as her sender key. When the ephemeral key and the
+sender key are the same, clients may indicate that a message is "intentionally
+anonymous" as opposed to "from an unknown sender".
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: I like these new lists, they're easier to follow.

File: go/saltpack/specs/saltpack_encryption.md [hunk 10]
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
...
+1. Generate a random 32-byte **payload key**.
+2. Generate a random ephemeral keypair, using
+   [`crypto_box_keypair`](http://nacl.cr.yp.to/box.html).
+3. Encrypt the sender's long-term public key using
+   [`crypto_secretbox`](http://nacl.cr.yp.to/secretbox.html) with the **payload
+   key**, to create the **sender secretbox**. (See [Nonces](#nonces).)
+4. For each recipient, encrypt the **payload key** using
+   [`crypto_box`](http://nacl.cr.yp.to/box.html) with the recipient's public
+   key and the ephemeral private key. (See [Nonces](#nonces).) Assemble these
+   into the **recipients list**.
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: Say to optionally include the receiver KID with each ciphertext?

File: go/saltpack/specs/saltpack_encryption.md [hunk 11]
```diff
@@ -113,30 +120,81 @@ A recipient pair is a two-element list:
...
+
+1. Deserialize the **header length** from the message stream using MessagePack.
+2. Read exactly **header length** bytes from the message stream.
+3. As above, hash the bytes from #2 to give the **header hash**.
+4. Deserialize the bytes from #2 using MessagePack to give the header list.
+5. Sanity check the **format name**, **version**, and **mode**.
+6. Precompute the ephemeral shared secret using
+   [`crypto_box_beforenm`](http://nacl.cr.yp.to/box.html) with the **ephemeral
+   public key** and the sender's private key.
+7. Try to open each of the **payload key boxes** in the recipients list using
```
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: As before, do you need to iterate if the your KID is explicitly listed?

**Comment by maxtaco** [comment]
made a few more comments on the new stuff....

File: go/saltpack/specs/saltpack_encryption.md (hunk 7 above)
Comment (oconnor663) [inline go/saltpack/specs/saltpack_encryption.md]: Every time, or just the first time?

File: go/saltpack/specs/saltpack_encryption.md (hunk 8 above)
Comment (oconnor663) [inline go/saltpack/specs/saltpack_encryption.md]: Woops!

File: go/saltpack/specs/saltpack_encryption.md (hunk 7 above)
Comment (maxtaco) [inline go/saltpack/specs/saltpack_encryption.md]: just the first. Sorry for the exceptional first case :)

File: go/saltpack/specs/saltpack_encryption.md (hunk 11 above)
Comment (oconnor663) [inline go/saltpack/specs/saltpack_encryption.md]: This one's clarified in the paragraph just below.

File: go/saltpack/specs/saltpack_encryption.md (hunk 7 above)
Comment (oconnor663) [inline go/saltpack/specs/saltpack_encryption.md]: IS THERE NO END TO YOUR MADNESS

**Comment by oconnor663** [comment]
Landed f2f2281a37083e238f50f462d30219dcc02a80a1.

**Comment by oconnor663** [comment]
Oops I mean 316ddaf5329bb73646f5d2df723c86d3705ae020. I got my branches confused.

**Pull request closed without merging**